script/dependencies/ProtegoDeploy.sol

```diff
@@ -0,0 +1,30 @@
+// SPDX-FileCopyrightText: © 2023 Dai Foundation <www.daifoundation.org>
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity ^0.8.16;
+
+import {ScriptTools} from "dss-test/ScriptTools.sol";
+import {Protego} from "src/Protego.sol";
+import {ProtegoInstance} from "./ProtegoInstance.sol";
+
+struct ProtegoDeployParams {
+    address pause;
+}
+
+library ProtegoDeploy {
```

Comment:
In this case, it might look like an overkill because it's very simple, but this approach makes it easier to test deploy/init scripts instead of writing everything as a Foundry script.
So the answer is that we want to be consistent with other repos. Maybe to keep consistency we might want to write some tests, even if it's trivial.
You can for instance check [`dss-lite-psm`](https://github.com/makerdao/dss-lite-psm/tree/341c533d604fb56be69a847d74e8b0a523119c9e/script/dependencies) for reference.

Reply:
For testing, Foundry actually recommends to directly ["test your scripts"](https://book.getfoundry.sh/tutorials/best-practices#:~:text=on%20your%20project.-,Test%20your%20scripts.,-Unit%20test%20them). But I think the main reason to use a library from a script is more than just code/organisation. From my understanding, a library would actually be deployed first in order to be used in a script, but then all of the operations in that library would be executed in a single transaction compared to a `s.sol` script where every operation is a separate transaction. In other words, it is to prevent front-running possibility. Although the above doesn't make any difference in this particular case, I agree with your point, let's keep it for consistency

Reply:
> For testing, Foundry actually recommends to directly ["test your scripts"](https://book.getfoundry.sh/tutorials/best-practices#:~:text=on%20your%20project.-,Test%20your%20scripts.,-Unit%20test%20them).

Yeah, but honestly it feels very cumbersome to test scripts because off all the context the `Script` base contract brings.

The script deals with "all things external", such as reading from config files/environment, while the library remains agnostic of configuration.

> [...] a library would actually be deployed first in order to be used in a script [...]

Actually, we don't deploy such libraries, they are imported as-is by spells. See [this example](https://github.com/makerdao/spells-mainnet/blob/master/archive/2023-07-14-DssSpell/dependencies/dss-flappers/FlapperInit.sol).

> [...] but then all of the operations in that library would be executed in a single transaction compared to a s.sol script where every operation is a separate transaction.

Since the library has only internal functions, the code is actually inlined by the compiler, so we actually have multiple transactions being executed when imported by a script.

Reply:
> Since the library has only internal functions, the code is actually inlined by the compiler

Thanks for pointing that out, I totally missed that fact.